extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/data/input/kafkainput/KafkaInputFormatTest.java

```diff
@@ -0,0 +1,329 @@
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+
+public class KafkaInputFormatTest
+{
+  private KafkaRecordEntity inputEntity;
+  private long timestamp = DateTimes.of("2021-06-24").getMillis();
+  private static final Iterable<Header> SAMPLE_HEADERS = ImmutableList.of(new Header() {
+      @Override
+      public String key()
+      {
+        return "encoding";
+      }
+      @Override
+      public byte[] value()
+      {
+        return "application/json".getBytes(StandardCharsets.UTF_8);
+      }
+    },
+      new Header() {
+      @Override
+      public String key()
+      {
+        return "kafkapkc";
+      }
+      @Override
+      public byte[] value()
+      {
+        return "pkc-bar".getBytes(StandardCharsets.UTF_8);
+      }
+    });
+  private KafkaInputFormat format;
+
+  @Before
+  public void setUp()
+  {
+    format = new KafkaInputFormat(
+        new KafkaStringHeaderFormat(null),
+        // Key Format
+        new JsonInputFormat(
+            new JSONPathSpec(true, ImmutableList.of()),
+            null, null, false //make sure JsonReader is used
+        ),
+        // Value Format
+        new JsonInputFormat(
+            new JSONPathSpec(
+                true,
+                ImmutableList.of(
+                    new JSONPathFieldSpec(JSONPathFieldType.ROOT, "root_baz", "baz"),
+                    new JSONPathFieldSpec(JSONPathFieldType.ROOT, "root_baz2", "baz2"),
+                    new JSONPathFieldSpec(JSONPathFieldType.PATH, "path_omg", "$.o.mg"),
+                    new JSONPathFieldSpec(JSONPathFieldType.PATH, "path_omg2", "$.o.mg2"),
+                    new JSONPathFieldSpec(JSONPathFieldType.JQ, "jq_omg", ".o.mg"),
+                    new JSONPathFieldSpec(JSONPathFieldType.JQ, "jq_omg2", ".o.mg2")
+                )
+            ),
+            null, null, false //make sure JsonReader is used
+        ),
+        "kafka.newheader.", "kafka.newkey.", "kafka.newts."
+    );
+  }
+
+  @Test
+  public void testWithHeaderKeyAndValue() throws IOException
+  {
+    final byte[] key = StringUtils.toUtf8(
+        "{\n"
+        + "    \"key\": \"sampleKey\"\n"
+        + "}");
+
+    final byte[] payload = StringUtils.toUtf8(
+        "{\n"
+        + "    \"timestamp\": \"2021-06-24\",\n"
+        + "    \"bar\": null,\n"
+        + "    \"foo\": \"x\",\n"
+        + "    \"baz\": 4,\n"
+        + "    \"o\": {\n"
+        + "        \"mg\": 1\n"
+        + "    }\n"
+        + "}");
+
+    Headers headers = new RecordHeaders(SAMPLE_HEADERS);
+    inputEntity = new KafkaRecordEntity(new ConsumerRecord<byte[], byte[]>(
+        "sample", 0, 0, timestamp,
```

Comment:
some older Kafka message formats do not have a timestamp (I think it's -1) we should figure out how we want to handle that